From: Max Lv <max.c.lv@gmail.com>
Date: Sat, 14 Oct 2017 07:21:08 +0800
Subject: Fix #1734

Fix potential local exploit issue.
Thanks to X41 D-Sec GmbH, Niklas Abel, for the reporting:
 https://www.x41-dsec.de/lab/advisories/x41-2017-010-shadowsocks-libev/

(cherry picked from commit c67d275803dc6ea22c558d06b1f7ba9f94cd8de3)
---
 src/manager.c | 24 ++++++++++++++++--------
 1 file changed, 16 insertions(+), 8 deletions(-)

diff --git a/src/manager.c b/src/manager.c
index 5672d2b..aee69fd 100644
--- a/src/manager.c
+++ b/src/manager.c
@@ -84,7 +84,7 @@ setnonblocking(int fd)
 }
 
 static void
-build_config(char *prefix, struct server *server)
+build_config(char *prefix, struct manager_ctx *manager, struct server *server)
 {
     char *path    = NULL;
     int path_size = strlen(prefix) + strlen(server->port) + 20;
@@ -100,9 +100,15 @@ build_config(char *prefix, struct server *server)
         return;
     }
     fprintf(f, "{\n");
-    fprintf(f, "\"server_port\":\"%s\",\n", server->port);
-    fprintf(f, "\"password\":\"%s\",\n", server->password);
-    fprintf(f, "}\n");
+    fprintf(f, "\"server_port\":%d,\n", atoi(server->port));
+    fprintf(f, "\"password\":\"%s\"", server->password);
+    if (manager->method)
+        fprintf(f, ",\n\"method\":\"%s\"", manager->method);
+    if (manager->fast_open)
+        fprintf(f, ",\n\"fast_open\": %d", manager->fast_open);
+    if (manager->mode)
+        fprintf(f, ",\n\"mode\":\"%d\"", manager->mode);
+    fprintf(f, "\n}\n");
     fclose(f);
     ss_free(path);
 }
@@ -112,14 +118,16 @@ construct_command_line(struct manager_ctx *manager, struct server *server)
 {
     static char cmd[BUF_SIZE];
     int i;
+    int port;
 
-    build_config(working_dir, server);
+    port = atoi(server->port);
+
+    build_config(working_dir, manager, server);
 
     memset(cmd, 0, BUF_SIZE);
     snprintf(cmd, BUF_SIZE,
-             "%s -m %s --manager-address %s -f %s/.shadowsocks_%s.pid -c %s/.shadowsocks_%s.conf",
-             executable, manager->method, manager->manager_address,
-             working_dir, server->port, working_dir, server->port);
+             "%s --manager-address %s -f %s/.shadowsocks_%d.pid -c %s/.shadowsocks_%d.conf",
+             executable, manager->manager_address, working_dir, port, working_dir, port);
 
     if (manager->acl != NULL) {
         int len = strlen(cmd);
